### PR TITLE
CP-24802: Add new constraint in Pool.join for the Host not having Pool_size feature

### PR DESCRIPTION
**Pull request by sharady**

### CODE REVIEW

File: ocaml/xapi/xapi_pool.ml [hunk 1]
```diff
@@ -111,6 +111,13 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
         raise (Api_errors.Server_error (code, ["The pool uses v6d. Pool edition list = " ^ pool_edn_list_str]))
   in
 
+  (* For an unlicensed pool, the maximum pool size allowed must be restricted to 3 hosts *)
+  let assert_pool_size_unrestricted () =
+    if (not (Pool_features.is_enabled ~__context Features.Pool_size)) &&
+      (List.length (Client.Host.get_all_records ~rpc ~session_id) >= Xapi_globs.unlicensed_pool_size_limit) then
+    raise (Api_errors.Server_error(Api_errors.license_restriction, []))
```
Comment (robhoes) [inline ocaml/xapi/xapi_pool.ml]: This should be indented one step further.

Comment (robhoes) [reply ocaml/xapi/xapi_pool.ml]: Also, the error needs a `"feature"` argument.

File: ocaml/xapi/xapi_pool.ml [hunk 2]
```diff
@@ -111,6 +111,13 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
         raise (Api_errors.Server_error (code, ["The pool uses v6d. Pool edition list = " ^ pool_edn_list_str]))
   in
 
+  (* For an unlicensed pool, the maximum pool size allowed must be restricted to 3 hosts *)
+  let assert_pool_size_unrestricted () =
+    if (not (Pool_features.is_enabled ~__context Features.Pool_size)) &&
+      (List.length (Client.Host.get_all_records ~rpc ~session_id) >= Xapi_globs.unlicensed_pool_size_limit) then
```
Comment (robhoes) [inline ocaml/xapi/xapi_pool.ml]: `get_all` is good enough, and more efficient.

File: ocaml/xapi/xapi_pool.ml [hunk 3]
```diff
@@ -433,6 +440,7 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
   assert_management_vlan_are_same ();
   assert_external_auth_matches ();
   assert_restrictions_match ();
+  assert_pool_size_unrestricted ();
```
Comment (robhoes) [inline ocaml/xapi/xapi_pool.ml]: Let's move this up in the list, because it's quick, so that we can fail early.

File: ocaml/xapi/xapi_pool.ml [hunk 4]
```diff
@@ -111,6 +111,13 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
         raise (Api_errors.Server_error (code, ["The pool uses v6d. Pool edition list = " ^ pool_edn_list_str]))
   in
 
+  (* The maximum pool size allowed must be restricted to 3 hosts for the pool which does not have Pool_size feature *)
+  let assert_pool_size_unrestricted () =
+    if (not (Pool_features.is_enabled ~__context Features.Pool_size)) &&
+      (List.length (Client.Host.get_all ~rpc ~session_id) >= Xapi_globs.restricted_pool_size) then
+      raise (Api_errors.Server_error(Api_errors.license_restriction, []))
```
Comment (robhoes) [inline ocaml/xapi/xapi_pool.ml]: Here, the error still needs a "feature" argument.

Comment (sharady) [reply ocaml/xapi/xapi_pool.ml]: Added, thanks for the review.

**Pull request merged**